### PR TITLE
Fix exception and enable event logging when service has no name

### LINKED ISSUE #62923: First chance exceptions and missing logging when a service isn't given a service name
See https://github.com/dotnet/runtime/issues/62579#issuecomment-995320972

The default behavior for ServiceBase when a ServiceName isn't specified is to use an empty string.  Per https://docs.microsoft.com/en-us/windows/win32/api/winsvc/ns-winsvc-service_table_entrya 
> If the service is installed with the SERVICE_WIN32_OWN_PROCESS service type, this member is ignored, but cannot be NULL. This member can be an empty string ("").

The problem, however, is that later on we use `ServiceName` to construct a log: 
https://github.com/dotnet/runtime/blob/1a83f61a5a3368250e5bed2baf28aed0f9e59508/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L238-L241

And EventLog will throw for an empty `Source`:
https://github.com/dotnet/runtime/blob/1a83f61a5a3368250e5bed2baf28aed0f9e59508/src/libraries/System.Diagnostics.EventLog/src/System/Diagnostics/EventLogInternal.cs#L1300-L1301

So in this common scenario all of these calls to WriteEventLog result in first chance exceptions, and nothing is logged:
https://github.com/dotnet/runtime/blob/1a83f61a5a3368250e5bed2baf28aed0f9e59508/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L944-L958

In this scenario we can actually acquire the service name from SCM, it's returned in the ServiceMain callback https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nc-winsvc-lpservice_main_functiona
> the first argument (lpServiceArgVectors[0]) is the name of the service

I have confirmed this to be the case.  So we can simply capture it here and set `ServiceName` rather than skipping it:
https://github.com/dotnet/runtime/blob/1a83f61a5a3368250e5bed2baf28aed0f9e59508/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L866-L880

### PR DESCRIPTION
**Pull request by danmoseley**
Fix https://github.com/dotnet/runtime/issues/62923

While all Windows services must have a non empty name, if you're starting a service in its own process, such that it's unambiguous which service you want to start, you do not need to pass that name. Therefore ServiceBase does not require you set ServiceName. However if ServiceBase is configured to write to the event log, which it is by default, it will fail because it uses the ServiceName as the source name. The result is an exception, and no expected messages, in the event log.

https://github.com/dotnet/runtime/blob/62b34d99c4d083c9f4ddcd715f19cdee195595ef/src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L260

The Microsoft.Extensions.Hosting library makes it easy to pass no service name to ServiceBase:

https://github.com/dotnet/runtime/blob/62b34d99c4d083c9f4ddcd715f19cdee195595ef/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetimeHostBuilderExtensions.cs#L31
https://github.com/dotnet/runtime/blob/62b34d99c4d083c9f4ddcd715f19cdee195595ef/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs#L35-L40
https://github.com/dotnet/runtime/blob/62b34d99c4d083c9f4ddcd715f19cdee195595ef/src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetimeOptions.cs#L13

The fix as indicated by @ericstj is for ServiceBase to populate its empty ServiceName using the value passed from the SCM in its callback.

- Increase timeouts that are sporadically being hit per test history data
- Add logging to TestServiceInstaller
- Update ServiceTypeOptions to include others present in winnt.h, but commented out
- Rename SERVICE_ constants to match winnt.h (it's less confusing)
- Remove defunct comment
- Add failing test
- Remove event log noise
- Make test pass

### CODE REVIEW

**Comment by danmoseley** [comment]
@dotnet/dnceng Helix error:
```

.packages/microsoft.dotnet.helix.sdk/7.0.0-beta.22179.1/tools/xharness-runner/XHarnessRunner.targets(122,5): error MSB4018: The "CreateXHarnessAppleWorkItems" task failed unexpectedly.
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
 ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: RevocationStatusUnknown
```

File: src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs [hunk 1]
```diff
@@ -65,25 +65,44 @@ internal static partial class ServiceOptions
...
             internal const int SERVICE_TYPE_ALL =
-                SERVICE_TYPE_WIN32 |
-                SERVICE_TYPE_ADAPTER |
-                SERVICE_TYPE_DRIVER |
-                SERVICE_TYPE_INTERACTIVE_PROCESS;
+                SERVICE_WIN32 |
+                SERVICE_ADAPTER |
+                SERVICE_DRIVER |
+                SERVICE_INTERACTIVE_PROCESS;
+              //SERVICE_USER_SERVICE |
```
Comment (jkotas) [inline src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs:103]: This definition should match the definition of `SERVICE_TYPE_ALL` in Windows SDK; or the name of this constant should be something else.

File: src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs [hunk 2]
```diff
@@ -174,17 +174,17 @@ private void DeleteService()
             using (var serviceManagerHandle = new SafeServiceHandle(Interop.Advapi32.OpenSCManager(null, null, Interop.Advapi32.ServiceControllerOptions.SC_MANAGER_ALL)))
             {
                 if (serviceManagerHandle.IsInvalid)
-                    throw new Win32Exception("Could not open SCM");
+                    throw new Win32Exception($"Could not open SCM. {new Win32Exception().Message}");
```
Comment (jkotas) [inline src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs:177]: `Win32Exception` constructor that takes string reads the error code using `Marshal.GetLastPInvokeError()`. 

Formatting of the error message runs a lot of code that can modify `Marshal.GetLastPInvokeError()`. It means the outer Win32Exception can end up with a bogus error code.

This should manage the error codes explicitly.

**Comment by premun** [comment]
@danmoseley the SSL problem is already fixed in Arcade but runtime hasn't consumed the .NET 7.0 Arcade yet:
https://github.com/dotnet/runtime/issues/67349

File: src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs (hunk 2 above)
Comment (danmoseley) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs:177]: Yup. BTW should we have an API that gets the message for an error code in a cleaner way? Eg a static on Win32Exception.

File: src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs [hunk 3]
```diff
@@ -174,17 +180,24 @@ private void DeleteService()
...
+                        string errorMessage = new Win32Exception().Message;
+                        throw new Win32Exception($"Could not find service '{ServiceName}'. {errorMessage}");
+                    }
 
                     TestService.DebugTrace("TestServiceInstaller: instructing ServiceController to Delete service " + ServiceName);
                     if (!Interop.Advapi32.DeleteService(serviceHandle))
                     {
-                        throw new Win32Exception($"Could not delete service '{ServiceName}'");
+                        string errorMessage = new Win32Exception().Message;
+                        throw new Win32Exception($"Could not delete service '{ServiceName}'. {errorMessage}");
```
Comment (jkotas) [inline src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs:200]: This ends up calling https://github.com/dotnet/runtime/blob/441271c95fe6e4d2b4bada7deb037a44d0eb8c74/src/libraries/System.Private.CoreLib/src/System/ComponentModel/Win32Exception.cs#L46 that will read the last PInvoke error that can be corrupted by now.

Comment (danmoseley) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs:200]: Ugh - Win32Exception is a bit of a trap. So is this the correct pattern to achieve this?
```c#
Win32Exception ex = new Win32Exception();
string errorMessage= ex.Message;
string code = ex.NativeErrorCode;
throw new Win32Exception(code, $"Could not delete service '{ServiceName}'. {errorMessage}");
```

Comment (jkotas) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs:200]: Yes, that is correct. I think writing it like this looks better:
```
int errorCode = Marshal.GetLastWin32Error();
string errorMessage = new Win32Exception(errorCode).Message;
throw new Win32Exception(errorCode, $"Could not delete service '{ServiceName}'. {errorMessage}");
```

Comment (danmoseley) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs:200]: similar
https://github.com/danmoseley/runtime/blob/a431bbc76dd80cb6f960725547ad3ccb5c9cf679/src/libraries/System.Drawing.Common/src/misc/DbgUtil.cs#L57-L59

I wonder what would be an appropriate new API to make this (seemingly reasonable) pattern easier.
```
throw Win32Exception.WithMessagePrefix($"Could not delete service '{ServiceName}.');
```
?

Comment (danmoseley) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs:200]: Actually, the above suggestion would still run string formatting machinery before reading the last error.

Comment (jkotas) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs]: I think an API that forces you to call `Marshal.GetLastWin32Error` yourself and gives you the error message for the OS error code would be more flexible.

I think the current behavior of Win32Exceptions that calls `Marshal.GetLastWin32Error` for you is very error prone.

Comment (danmoseley) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs]: Hmm, so eg this?  (assuming I want the code in the exception object too)
```c#
int errorCode = Marshal.GetLastPinvokeError();
int errorMessage = Marshal.GetMessageForPinvokeError(errorCode);
throw new Win32Exception(errorCode, $"Could not delete service '{ServiceName}'. {errorMessage}");
```

Comment (danmoseley) [reply src/libraries/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestServiceInstaller.cs]: created #67872

**Comment by danmoseley** [comment]
Any other feedback?

File: src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs [hunk 4]
```diff
@@ -65,25 +65,44 @@ internal static partial class ServiceOptions
...
+                SERVICE_RECOGNIZER_DRIVER;
+
+            internal const int SERVICE_WIN32_OWN_PROCESS = 0x00000010;
+            internal const int SERVICE_WIN32_SHARE_PROCESS = 0x00000020;
+
+            internal const int SERVICE_WIN32 =
+                SERVICE_WIN32_OWN_PROCESS |
+                SERVICE_WIN32_SHARE_PROCESS;
+
+            //internal const int SERVICE_USER_SERVICE = 0x00000040;
```
Comment (jkotas) [inline src/libraries/Common/src/Interop/Windows/Advapi32/Interop.ServiceProcessOptions.cs:85]: These can be deleted if they are not used. We only keep the interop definitions that are actually used.

File: src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs [hunk 5]
```diff
@@ -887,8 +873,13 @@ public unsafe void ServiceMainCallback(int argCount, IntPtr argPointer)
                 {
                     char** argsAsPtr = (char**)argPointer.ToPointer();
```
Comment (jkotas) [inline src/libraries/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs:874]:
```suggestion
                    char** argsAsPtr = (char**)argPointer;
```
unnecessary

**Review by jkotas — APPROVED** [review]

**Comment by danmoseley** [comment]
Interpreter failure is #67880

**Review by ericstj — COMMENTED** [review]
LGTM, thank you for fixing this!

**Pull request merged**